File: internal/store/outreach_sources.go
package store

import "strings"

// OutreachSource is one brain page bound to a knowledge need (experience/voice)
// for the outreach + answers pipeline, with its whole text cached locally.
type OutreachSource struct {
	Need       string `json:"need"`
	PageID     string `json:"page_id"`
	Title      string `json:"title"`
	Content    string `json:"content"`
	Version    string `json:"version"`
	ResolvedAt string `json:"resolved_at"`
}

// ListOutreachSources returns every cached source, ordered by need then title.
func (db *DB) ListOutreachSources() ([]OutreachSource, error) {
	const q = `SELECT need, page_id, title, content, version, resolved_at
FROM outreach_sources ORDER BY need, title`
	rows, err := db.Query(q)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []OutreachSource
	for rows.Next() {
		var s OutreachSource
		if err := rows.Scan(&s.Need, &s.PageID, &s.Title, &s.Content, &s.Version, &s.ResolvedAt); err != nil {
			return nil, err
		}
		out = append(out, s)
	}
	return out, rows.Err()
}

// OutreachKnowledge concatenates the cached text of every page bound to a need
// — the whole-fetched bundle the fill step and honesty checker reason over.
// Empty string means the need has no resolved sources.
func (db *DB) OutreachKnowledge(need string) (string, error) {
	const q = `SELECT title, content FROM outreach_sources WHERE need = ? ORDER BY title`
	rows, err := db.Query(q, need)
	if err != nil {
		return "", err
	}
	defer rows.Close()
	var parts []string
	for rows.Next() {
		var title, content string
		if err := rows.Scan(&title, &content); err != nil {
			return "", err
		}
		if strings.TrimSpace(content) == "" {
			continue
		}
		if title != "" {
			parts = append(parts, "# "+title+"\n\n"+content)
		} else {
			parts = append(parts, content)
		}
	}
	if err := rows.Err(); err != nil {
		return "", err
	}
	return strings.Join(parts, "\n\n---\n\n"), nil
}

// ReplaceOutreachSources swaps the cached set for one need in a transaction
// (delete-all-for-need + insert) — the discovery/refresh write.
func (db *DB) ReplaceOutreachSources(need string, sources []OutreachSource) error {
	tx, err := db.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()
	if _, err := tx.Exec(`DELETE FROM outreach_sources WHERE need = ?`, need); err != nil {
		return err
	}
	const ins = `INSERT INTO outreach_sources (need, page_id, title, content, version)
VALUES (?, ?, ?, ?, ?)`
	for _, s := range sources {
		if _, err := tx.Exec(ins, need, s.PageID, s.Title, s.Content, s.Version); err != nil {
			return err
		}
	}
	return tx.Commit()
}

// UpsertOutreachSource adds or refreshes one (need, page_id) row. Used by tests
// to seed a cached bundle; the live path replaces whole need-sets via discovery
// (ReplaceOutreachSources).
func (db *DB) UpsertOutreachSource(s OutreachSource) error {
	const q = `
INSERT INTO outreach_sources (need, page_id, title, content, version, resolved_at)
VALUES (?, ?, ?, ?, ?, CURRENT_TIMESTAMP)
ON CONFLICT(need, page_id) DO UPDATE SET
    title = excluded.title, content = excluded.content,
    version = excluded.version, resolved_at = CURRENT_TIMESTAMP`
	_, err := db.Exec(q, s.Need, s.PageID, s.Title, s.Content, s.Version)
	return err
}

File: scout/store/outreach_sources.py
"""The outreach knowledge store: prose docs bound to knowledge needs.

Each row carries an origin — 'brain' for a page the sync whole-fetched, 'local'
for the doc typed in Settings → Knowledge (page_id 'local', at most one per
need). The one rule: a sync may overwrite brain rows and never touches local
ones. Readers concatenate both, so every consumer gets typed knowledge for free.
"""

from __future__ import annotations

import sqlite3
from dataclasses import dataclass

from ._helpers import tx

ORIGIN_BRAIN = "brain"
ORIGIN_LOCAL = "local"
LOCAL_PAGE_ID = "local"


@dataclass
class OutreachSource:
    need: str = ""
    page_id: str = ""
    title: str = ""
    content: str = ""
    version: str = ""
    resolved_at: str = ""
    origin: str = ORIGIN_BRAIN


def list_outreach_sources(con: sqlite3.Connection) -> list[OutreachSource]:
    """Every source, ordered by need, then local first, then title."""
    rows = con.execute(
        "SELECT need, page_id, title, content, version, resolved_at, origin "
        "FROM outreach_sources ORDER BY need, origin = 'local' DESC, title"
    ).fetchall()
    return [
        OutreachSource(
            need=r[0],
            page_id=r[1],
            title=r[2],
            content=r[3],
            version=r[4],
            resolved_at=r[5],
            origin=r[6],
        )
        for r in rows
    ]


def outreach_knowledge(con: sqlite3.Connection, need: str) -> str:
    """Concatenate the text of every doc bound to a need — the typed doc first,
    then the brain pages. Empty string means the need has nothing on file."""
    rows = con.execute(
        "SELECT title, content FROM outreach_sources WHERE need = ? "
        "ORDER BY origin = 'local' DESC, title",
        (need,),
    ).fetchall()
    parts: list[str] = []
    for title, content in rows:
        if content.strip() == "":
            continue
        if title != "":
            parts.append(f"# {title}\n\n{content}")
        else:
            parts.append(content)
    return "\n\n---\n\n".join(parts)


def replace_outreach_sources(
    con: sqlite3.Connection, need: str, sources: list[OutreachSource]
) -> None:
    """Swap the brain-synced set for one need in a transaction (delete-all +
    insert). Only brain rows are replaced — a local row survives every sync."""
    with tx(con):
        con.execute(
            "DELETE FROM outreach_sources WHERE need = ? AND origin = ?", (need, ORIGIN_BRAIN)
        )
        for s in sources:
            con.execute(
                "INSERT INTO outreach_sources (need, page_id, title, content, version, origin) "
                "VALUES (?, ?, ?, ?, ?, ?)",
                (need, s.page_id, s.title, s.content, s.version, ORIGIN_BRAIN),
            )


def upsert_outreach_source(con: sqlite3.Connection, s: OutreachSource) -> None:
    """Add or refresh one (need, page_id) row."""
    con.execute(
        """
INSERT INTO outreach_sources (need, page_id, title, content, version, origin, resolved_at)
VALUES (?, ?, ?, ?, ?, ?, CURRENT_TIMESTAMP)
ON CONFLICT(need, page_id) DO UPDATE SET
    title = excluded.title, content = excluded.content, version = excluded.version,
    origin = excluded.origin, resolved_at = CURRENT_TIMESTAMP""",
        (s.need, s.page_id, s.title, s.content, s.version, s.origin or ORIGIN_BRAIN),
    )


def get_local_source(con: sqlite3.Connection, need: str) -> str:
    """The typed doc for a need, or "" when none has been saved."""
    row = con.execute(
        "SELECT content FROM outreach_sources WHERE need = ? AND page_id = ?",
        (need, LOCAL_PAGE_ID),
    ).fetchone()
    return row[0] if row is not None else ""


def put_local_source(con: sqlite3.Connection, need: str, content: str) -> None:
    """Save the typed doc for a need. Blank content removes the row, so "nothing
    typed" and "cleared" look the same to every reader."""
    if content.strip() == "":
        con.execute(
            "DELETE FROM outreach_sources WHERE need = ? AND page_id = ?", (need, LOCAL_PAGE_ID)
        )
        return
    upsert_outreach_source(
        con,
        OutreachSource(need=need, page_id=LOCAL_PAGE_ID, content=content, origin=ORIGIN_LOCAL),
    )
